### ISSUE TITLE
[Bug] Mark navigation really slow when the mark quick-fix list is open

### ISSUE THREAD

**Issue by pidgeon777** [LeonHeidelbach/trailblazer.nvim#38]
### Did you check docs and existing issues?

- [X] I have read all the trailblazer.nvim docs
- [X] I have searched the existing issues of trailblazer.nvim
- [X] I have searched the exsiting issues of plugins related to this issue

### Neovim version (nvim -v)

v0.8.3

### Operating system/version

Windows 10 21H2

### Describe the bug

When the mark quick-fix list is open, marks navigation with the `TrailBlazerPeekMovePreviousUp` and `TrailBlazerPeekMoveNextDown` commands is really slow, compared to when the quick-fix mark list is closed.

### Steps To Reproduce

1. Start a blank `Neovim` session.
2. Open 10 buffers containing each many lines of code (e.g. LUA, C++, etc.).

> NOTE: In my test, I used a number of buffers other than 10.

3. Create a single mark for each buffer, so that it will be possible to switch between all of those buffers with the `TrailBlazerPeekMovePreviousUp` and `TrailBlazerPeekMoveNextDown` commands.

Two tests:

* Test A : with mark list quick-fix closed

1. Position the cursor on the first mark starting from the top list position.
2. Start profiling with:

```vim
:profile start profile.log
:profile func *
:profile file *
```

3. Execute `TrailBlazerPeekMoveNextDown` 9 times so that the cursor is positioned on the last mark.
4. Stop profiling with:

```vim
:profile stop
```

My output:

https://gist.github.com/pidgeon777/d1bea4a7955605a963f5f14fcb88deef

* Test B : with mark list quick-fix open

1. Open the quick-fix panel with `TrailBlazerToggleTrailMarkList`.
2. Position the cursor on the first mark starting from the top list position.
3. Start profiling with:

```vim
:profile start profile.log
:profile func *
:profile file *
```

4. Execute `TrailBlazerPeekMoveNextDown` 9 times so that the cursor is positioned on the last mark.
5. Stop profiling with:

```vim
:profile stop
```

My output:

https://gist.github.com/pidgeon777/67202fa42d33a9d9361d2feb076477a2

* Observations:

Timings of **Test A**:

```log
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    5   0.113866   0.112961  <SNR>1_LoadFTPlugin()
    5   0.062757             <SNR>11_SynSet()
    5   0.030490             <SNR>2_LoadIndent()
   32   0.018744   0.018509  <SNR>61_display_annotation()
   47   0.018551   0.010289  <SNR>60_SetupAutoCommands()
   71   0.008575   0.007703  <SNR>16_Highlight_Matching_Pair()
   42   0.005942   0.003510  <SNR>60_HighlightEOLWhitespace()
   86   0.002365             <SNR>60_ClearHighlighting()
   89   0.002176             <SNR>60_ShouldHighlight()
   75   0.000915             <SNR>16_Remove_Matches()
   47   0.000876             <SNR>60_ShouldStripWhitespaceOnSave()
   30   0.000236             bm#has_bookmark_at_line()
    4   0.000014             <SNR>36_dopopd()
                             <SNR>111_setup_toc()
                             <SNR>80_try_cmd()
                             9()
                             provider#clipboard#Call()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    5   0.113866   0.112961  <SNR>1_LoadFTPlugin()
    5              0.062757  <SNR>11_SynSet()
    5              0.030490  <SNR>2_LoadIndent()
   32   0.018744   0.018509  <SNR>61_display_annotation()
   47   0.018551   0.010289  <SNR>60_SetupAutoCommands()
   71   0.008575   0.007703  <SNR>16_Highlight_Matching_Pair()
   42   0.005942   0.003510  <SNR>60_HighlightEOLWhitespace()
   86              0.002365  <SNR>60_ClearHighlighting()
   89              0.002176  <SNR>60_ShouldHighlight()
   75              0.000915  <SNR>16_Remove_Matches()
   47              0.000876  <SNR>60_ShouldStripWhitespaceOnSave()
   30              0.000236  bm#has_bookmark_at_line()
    4              0.000014  <SNR>36_dopopd()
                             <SNR>80_try_cmd()
                             9()
                             provider#clipboard#Call()
                             <SNR>111_setup_toc()
```

Timings of **Test B**:

```log
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
  245   7.050313   6.613126  <SNR>1_LoadFTPlugin()
  245   3.849984   3.825071  <SNR>11_SynSet()
  245   1.672516             <SNR>2_LoadIndent()
  485   0.114609   0.075307  <SNR>60_SetupAutoCommands()
  485   0.016504             <SNR>60_ClearHighlighting()
  525   0.011392             <SNR>60_ShouldHighlight()
  485   0.008247             <SNR>60_ShouldStripWhitespaceOnSave()
   43   0.004787   0.004323  <SNR>16_Highlight_Matching_Pair()
   40   0.004172   0.003159  <SNR>60_HighlightEOLWhitespace()
    7   0.002136   0.002108  <SNR>61_display_annotation()
   47   0.000514             <SNR>16_Remove_Matches()
    5   0.000028             bm#has_bookmark_at_line()
    4   0.000017             <SNR>36_dopopd()
    1   0.000012             <SNR>111_setup_toc()
                             <SNR>80_try_cmd()
                             9()
                             provider#clipboard#Call()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
  245   7.050313   6.613126  <SNR>1_LoadFTPlugin()
  245   3.849984   3.825071  <SNR>11_SynSet()
  245              1.672516  <SNR>2_LoadIndent()
  485   0.114609   0.075307  <SNR>60_SetupAutoCommands()
  485              0.016504  <SNR>60_ClearHighlighting()
  525              0.011392  <SNR>60_ShouldHighlight()
  485              0.008247  <SNR>60_ShouldStripWhitespaceOnSave()
   43   0.004787   0.004323  <SNR>16_Highlight_Matching_Pair()
   40   0.004172   0.003159  <SNR>60_HighlightEOLWhitespace()
    7   0.002136   0.002108  <SNR>61_display_annotation()
   47              0.000514  <SNR>16_Remove_Matches()
    5              0.000028  bm#has_bookmark_at_line()
    4              0.000017  <SNR>36_dopopd()
    1              0.000012  <SNR>111_setup_toc()
                             9()
                             provider#clipboard#Call()
                             <SNR>80_try_cmd()
```

We can notice how with the mark list quick-fix window open, there are big differences on how many times the following functions are called (i.e. function call ratio between **Test A** and **Test B**):

FUNCTION | Test A | Test B
---|---|---
`<SNR>1_LoadFTPlugin()` | 5 | 245
`<SNR>11_SynSet()` | 5 | 245
`<SNR>2_LoadIndent()` | 5 | 245
`<SNR>60_SetupAutoCommands()` | 47 | 485
`<SNR>60_ClearHighlighting()` | 86 | 485
`<SNR>60_ShouldHighlight()` | 89 | 525
`<SNR>60_ShouldStripWhitespaceOnSave()` | 47 | 485

With **Test B**, buffer switching was much more slower than with **Test A**, because some "heavy" functions (e.g. syntax and indent apply) where called many more times.

Interestingly, the following functions were called more times in **Test A** rather than in **Test B**, but those where "light":

FUNCTION | Test A | Test B
---|---|---
`<SNR>16_Highlight_Matching_Pair()` | 71 | 43
`<SNR>16_Remove_Matches()` | 75 | 47
`<SNR>61_display_annotation()` | 32 | 7
`bm#has_bookmark_at_line()` | 30 | 5

With the rest of the functions, there are no big differences between **Test A** and **Test B**:

FUNCTION | Test A | Test B
---|---|---
`<SNR>60_HighlightEOLWhitespace()` | 42 | 40
`<SNR>36_dopopd()` | 4 | 4
`<SNR>111_setup_toc()` | 1 | 1
`9()` | 1 | 1
`provider#clipboard#Call()` | 1 | 1
`<SNR>80_try_cmd()` | 1 | 1

It is important to note that a further **Test C** was performed, where the *raw* (not `trailblazer.nvim`) quick-fix window was populated simply with the buffer list (thus not using the `TrailBlazerToggleTrailMarkList` command), and buffer switching occurred with the `:cnext` commands.

The results were very similar to **Test A** (mark navigation with no quick-fix mark list open) in terms of timings and function calls:

https://gist.github.com/pidgeon777/3e355a42af4b9b928cb4807d2202772c

Timings of **Test C**:

```log
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
    5   0.119966   0.118921  <SNR>1_LoadFTPlugin()
    5   0.064960             <SNR>11_SynSet()
    5   0.028906             <SNR>2_LoadIndent()
   47   0.017287   0.009662  <SNR>60_SetupAutoCommands()
   28   0.014631   0.014461  <SNR>61_display_annotation()
   59   0.006307   0.005703  <SNR>16_Highlight_Matching_Pair()
   42   0.005532   0.003303  <SNR>60_HighlightEOLWhitespace()
   86   0.002173             <SNR>60_ClearHighlighting()
   89   0.002052             <SNR>60_ShouldHighlight()
   47   0.000766             <SNR>60_ShouldStripWhitespaceOnSave()
   63   0.000651             <SNR>16_Remove_Matches()
   25   0.000170             bm#has_bookmark_at_line()
    4   0.000015             <SNR>36_dopopd()
                             <SNR>80_try_cmd()
                             9()
                             provider#clipboard#Call()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
    5   0.119966   0.118921  <SNR>1_LoadFTPlugin()
    5              0.064960  <SNR>11_SynSet()
    5              0.028906  <SNR>2_LoadIndent()
   28   0.014631   0.014461  <SNR>61_display_annotation()
   47   0.017287   0.009662  <SNR>60_SetupAutoCommands()
   59   0.006307   0.005703  <SNR>16_Highlight_Matching_Pair()
   42   0.005532   0.003303  <SNR>60_HighlightEOLWhitespace()
   86              0.002173  <SNR>60_ClearHighlighting()
   89              0.002052  <SNR>60_ShouldHighlight()
   47              0.000766  <SNR>60_ShouldStripWhitespaceOnSave()
   63              0.000651  <SNR>16_Remove_Matches()
   25              0.000170  bm#has_bookmark_at_line()
    4              0.000015  <SNR>36_dopopd()
                             9()
                             provider#clipboard#Call()
                             <SNR>80_try_cmd()
```

Comparison of **Test A** and **Test C**:

FUNCTION | Test A | Test C
---|---|---
`<SNR>1_LoadFTPlugin()` | 5 | 5
`<SNR>11_SynSet()` | 5 | 5
`<SNR>2_LoadIndent()` | 5 | 5
`<SNR>60_SetupAutoCommands()` | 47 | 47
`<SNR>61_display_annotation()` | 32 | 28
`<SNR>16_Highlight_Matching_Pair()` | 71 | 59
`<SNR>60_HighlightEOLWhitespace()` | 42 | 42
`<SNR>60_ClearHighlighting()` | 86 | 86
`<SNR>60_ShouldHighlight()` | 89 | 89
`<SNR>60_ShouldStripWhitespaceOnSave()` | 47 | 47
`<SNR>16_Remove_Matches()` | 75 | 63

We can notice how results are very similar between **Test A** and **Test C**.

This lead me to the conclusion that the extreme slowness of **Test B** (mark navigation with mark list panel open) doesn't come from the fact that the quick-fix window used to display the list of marks is open itself (otherwise we would have had slowness even for **Test A**), but rather from the fact that the `trailblazer.nvim` plugin seems to always execute some heavy commands when the mark quick-fix list is open and a mark is jumped to.

As if the plugin didn't take into account the fact that a mark can reside in the same buffer as the current one, and that therefore it is not needed to do certain buffer operations again, like applying syntax, and so on (is happening, for example, a big number of times with the `1_LoadFTPlugin()` or `11_SynSet()` functions in **Test B**, and which cause the slowness).

Question is: given that buffer switching is very smooth with the *raw* quick-fix open and the `:cnext` command, what does cause the mark navigation slowness with the `trailblazer.vim` quick-fix mark list open, and why this slowness doesn't occur when the list is closed?

### Expected Behavior

See the above considerations.

**Comment by LeonHeidelbach** [comment]
After doing some more research and looking for different ways to update quickfix lists I could only find my currently implemented solution as being the "generally accepted" way to update quickfix lists. The only factor that is making this operation slow is hat TreeSitter tries to re-highlight the quickfix buffer text on every update, as your extensive testing shows.

**Comment by pidgeon777** [comment]
Yes, what I mean is that if the list didn't change, I would expect the plugin not to perform any action like reloading syntaxes, but just select the previous and next quick-fix item and navigate to the position, as it happens with `cnext` and `cprevious`.

The fact is that when executing those two commands when the mark quick-fix panel is open, then mark navigation occurs and is fast as using the *raw* quick-fix, but at some points, the navigation gets stuck eventually entering a sort of loop if continuing to execute for example the "navigate" up command, and it is not possible to reach all of the elements of the list.

To replicate this, it is enough to open some buffers, add some marks, open the mark quick-fix panel, and then execute the `cnext` and `cprevious` commands. As a result, the mark "pointer" in the quick-fix panel will get stuck at a certain point, eventually entering a "loop" (going back to a previously navigated mark).

If there would be a way to navigate the mark quick-fix panel using the *raw* quick-fix commands `cnext` and `cprevious`, then it would be a great temporary workaround to fix the navigation slowness.

**Comment by LeonHeidelbach** [comment]
After some testing, I definitely do notice quite the performance improvement with this change. I think this should be the solution.

**Comment by pidgeon777** [comment]
In the meanwhile, I found a bug. Configuration:

```lua
current_trail_mark_mode = "global_fpath_line_sorted"
move_to_nearest_before_peek = true
move_to_nearest_before_peek_motion_directive_down = "fpath_down"
move_to_nearest_before_peek_dist_type = "lin_char_dist"
```

Also map the `TrailBlazerPeekMoveNextDown` to a key, for example <M-Down> just for simplicity.

To replicate it is necessary to move to the last item of the trail mark quick-fix list. Once it is reached, keep the key pressed so that the `TrailBlazerPeekMoveNextDown` is continuously executed, holding the key down for example for 10 seconds.

As a result, after releasing the key, `Neovim` will be stuck for a few seconds before becoming active again, like something is running in the background. And this is what runs in the background while you hold down the key:

```log
FUNCTIONS SORTED ON TOTAL TIME
count  total (s)   self (s)  function
  295   7.687625   7.125271  <SNR>1_LoadFTPlugin()
  295   4.144193   4.117690  <SNR>11_SynSet()
  295   1.762824             <SNR>2_LoadIndent()
  585   0.086375   0.047285  <SNR>60_SetupAutoCommands()
  585   0.018898             <SNR>60_ClearHighlighting()
  587   0.011132             <SNR>60_ShouldHighlight()
  585   0.008988             <SNR>60_ShouldStripWhitespaceOnSave()
   16   0.001443   0.001278  <SNR>16_Highlight_Matching_Pair()
    4   0.000666   0.000655  <SNR>61_display_annotation()
   20   0.000216             <SNR>16_Remove_Matches()
    2   0.000114   0.000071  <SNR>60_HighlightEOLWhitespace()
    1   0.000023             <SNR>119_setup_toc()
    4   0.000018             <SNR>36_dopopd()
    2   0.000010             bm#has_bookmark_at_line()

FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
  295   7.687625   7.125271  <SNR>1_LoadFTPlugin()
  295   4.144193   4.117690  <SNR>11_SynSet()
  295              1.762824  <SNR>2_LoadIndent()
  585   0.086375   0.047285  <SNR>60_SetupAutoCommands()
  585              0.018898  <SNR>60_ClearHighlighting()
  587              0.011132  <SNR>60_ShouldHighlight()
  585              0.008988  <SNR>60_ShouldStripWhitespaceOnSave()
   16   0.001443   0.001278  <SNR>16_Highlight_Matching_Pair()
    4   0.000666   0.000655  <SNR>61_display_annotation()
   20              0.000216  <SNR>16_Remove_Matches()
    2   0.000114   0.000071  <SNR>60_HighlightEOLWhitespace()
    1              0.000023  <SNR>119_setup_toc()
    4              0.000018  <SNR>36_dopopd()
    2              0.000010  bm#has_bookmark_at_line()
```

I would expect no functions to be executed in the case that the last mark quick-fix item is reached and the `TrailBlazerPeekMoveNextDown` is executed.

So, what I mean, is that maybe this could be fixed as follows: when the last item of the quick-fix list is reached, do nothing when trying to navigate to the next item for example with the `TrailBlazerPeekMoveNextDown` command, because the current item is the last of the list.

It is important to note that the same issue occurs for the `TrailBlazerPeekMovePreviousUp` command after reaching the top of the mark list.

**Comment by LeonHeidelbach** [comment]
> I confirm the speed improvement.
> 
> I wonder if there could be still room for improvement. I suppose the final one would be to navigate the mark quick-fix only using the `cnext` and `cprevious` commands someway?

@pidgeon777 This is actually the way I implemented it now. As long as the contents of the current trail mark stack don't change, navigation through the quickfix list is done by calling `cnext` and `cprevious` if the current cursor index changes within the trail mark stack. If the cursor index does not change but the `update_trail_mark_list` function is called, the layout will be recomputed. That is why the highlight functions are called as soon as you reach the top or the bottom of the list and proceed to move in the respective direction. I guess this could be improved by adding a separate argument to the update function that signals if a motion was triggered where only `cnext` and `cprevious` calls are required.

I am currently quite busy though, so I will postpone futher performance improvement investigations to a later date. But I will definitely come back to this.

**Comment by pidgeon777** [comment]
> I guess this could be improved by adding a separate argument to the update function that signals if a motion was triggered where only cnext and cprevious calls are required.

Seems to be a good idea.

> I am currently quite busy though, so I will postpone futher performance improvement investigations to a later date. But I will definitely come back to this.

👍 in the meanwhile I'll report any found issues.

**Comment by pidgeon777** [comment]
Maybe it could be a good idea to reopen temporarily this issue?

**Comment by LeonHeidelbach** [comment]
@pidgeon777 I have this on my TODO list and see to it in the next couple of days. The initial problem from this issue was fixed, so no need to re-open it.

**Comment by LeonHeidelbach** [comment]
> I would expect no functions to be executed in the case that the last mark quick-fix item is reached and the `TrailBlazerPeekMoveNextDown` is executed.
> 
> So, what I mean, is that maybe this could be fixed as follows: when the last item of the quick-fix list is reached, do nothing when trying to navigate to the next item for example with the `TrailBlazerPeekMoveNextDown` command, because the current item is the last of the list.
> 
> It is important to note that the same issue occurs for the `TrailBlazerPeekMovePreviousUp` command after reaching the top of the mark list.

@pidgeon777 This should now be fixed e10e898535fa64c9b77d5ab5427c8f8fd4c680c4.

**Comment by pidgeon777** [comment]
I confirm this fixed the boundary issue. Great work!
